File: backend/app/services/alergia_service.py
from app.domain.models import Alergia
from app.domain.schemas import AlergiaCreate, AlergiaUpdate, AlergiaRead
from app.repositories import AlergiaRepository

class AlergiaService:
    def __init__(self, repo: AlergiaRepository):
        self.repo = repo

    def obtener_todas(self) -> list[AlergiaRead]:
        alergias = self.repo.get_all()
        return [AlergiaRead.model_validate(a) for a in alergias]
    
    def obtener_por_id(self, id: int) -> AlergiaRead | None:
        alergia = self.repo.get_by_id(id)
        if not alergia:
            raise ValueError(f"No se encontró la alergia con ID {id}.")
        return AlergiaRead.model_validate(alergia)
    
    def obtener_modelo_por_id(self, id: int) -> Alergia | None:
        alergia = self.repo.get_by_id(id)
        if not alergia:
            raise ValueError(f"No se encontró la alergia con ID {id}.")
        return alergia

    def crear_alergia(self, alergia_in: AlergiaCreate) -> AlergiaRead:
        alergia = Alergia.model_validate(alergia_in)
        return AlergiaRead.model_validate(self.repo.add(alergia))
    
    def actualizar(self, id: int, data: AlergiaUpdate) -> Alergia | None:
        alergia_actual = self.repo.get_by_id(id)
        if not alergia_actual:
            return None
        
        datos_limpios = data.model_dump(exclude_unset=True)
        return self.repo.update(alergia_actual, datos_limpios)

File: frontend/turnero-medico/src/service/alergiaService.ts
import axios from 'axios';
import {
  type Alergia,
  type AlergiaCreate,
  type AlergiaUpdate,
} from '../types/Alergia';

const API_URL = 'http://localhost:8000/alergias';

export class AlergiaService {
  static async getAll(): Promise<Alergia[]> {
    const res = await axios.get(API_URL + '/');
    return res.data;
  }

  static async getById(id: number): Promise<Alergia> {
    const res = await axios.get(API_URL + `/${id}`);
    return res.data;
  }

  static async create(data: AlergiaCreate): Promise<Alergia> {
    const res = await axios.post(API_URL + '/', data);
    return res.data;
  }

  static async update(id: number, data: AlergiaUpdate): Promise<Alergia> {
    const res = await axios.put(API_URL + `/${id}`, data);
    return res.data;
  }

  static async delete(id: number): Promise<void> {
    await axios.delete(API_URL + `/${id}`);
  }
}
